### PR TITLE
Sonar integration with Java Tests code coverage

### PR DESCRIPTION
**Pull request by pazone**
## What does this PR do?

Enables sonarqube code scanning and coverage reports for logstash-core. 
[Sonar project for logstash](https://sonar.elastic.dev/dashboard?id=elastic_logstash_AYm_nEbQaV3I-igkX1q9) is already created.

This configuration provides:

**Pull request analysis**: Analyses newly added code for a particular pull request
SonarQube automatically adds a comment with digest information. If no java files were changed in a pull request we will see “No coverage/duplication information” labels

**Branch analysis**: Runs on branches like main and analyzes the entire codebase. Associates received code coverage with the branch

## Why is it important?

Continuous static analysis and solid code coverage help us maintain well-designed and reliable code.

Furthermore we can setup desired quality gates for our [logstash sonar project](https://sonar.elastic.dev/dashboard?id=elastic_logstash_AYm_nEbQaV3I-igkX1q9), i.e. minimal coverage threshold for the new code and for branches analysis. 

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files (and/or docker env variables)
- [ ] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [ ] Sonarqube provides code coverage information on pull requests
- [ ] Sonarqube dashboard is properly setup and clearly represents branches and PR state
- [ ] Quality gate is set up and aligned with the team
- [ ] Quality profile is set up and aligned with the team

## How to test this PR locally

```
SONAR_TOKEN=$(vault read -field=token secret/ci/elastic-logstash/sonar-creds)
./gradlew javaTests jacocoTestReport sonar -Dsonar.token="${SONAR_TOKEN}" --console=plain --warning-mode all
```
## Related issues

- Closes https://github.com/elastic/ingest-dev/issues/2280

### CODE REVIEW

File: ci/docker_unit_tests.sh [hunk 1]
```diff
@@ -1,2 +1,3 @@
 #!/bin/bash
-ci/docker_run.sh logstash-unit-tests ci/unit_tests.sh $@
+SONAR_TOKEN=$(vault read -field=token secret/ci/elastic-logstash/sonar-creds)
```
Comment (robbavey) [inline ci/docker_unit_tests.sh:2]: Judging by the test failures, it looks like there is an issue pointing to the correct vault:

```
09:08:37 Error reading secret/ci/elastic-logstash/sonar-creds: Get "https://127.0.0.1:8200/v1/secret/ci/elastic-logstash/sonar-creds": dial tcp 127.0.0.1:8200: connect: connection refused
``

I think you may need to pass the vault environment variables into the jenkins jobs, probably doing something along the lines of https://github.com/elastic/infra/blob/9e0feddfb4d7409342734de4c8f247ea5c96b6cb/ci/jjb/logstash-ci/defs/elastic-logstash/on-demand.yml#L34 for the Logstash jobs defined [here](https://github.com/elastic/infra/tree/9e0feddfb4d7409342734de4c8f247ea5c96b6cb/ci/jjb/logstash-ci/defs/elastic-logstash)

**Review by robbavey — CHANGES_REQUESTED** [review]
I think there is more prep work needed in the Jenkins jobs to get this up and running - the ci failure is related to that

Also running `ci/unit_tests.sh` is broken due to a duplicate `console` definition

File: gradle.properties [hunk 2]
```diff
@@ -5,3 +5,8 @@ org.gradle.daemon=false
 systemProp.org.gradle.internal.http.connectionTimeout=120000
 systemProp.org.gradle.internal.http.socketTimeout=120000
 systemProp.org.gradle.internal.http.idleConnectionTimeout=120000
+
+# Sonar
+systemProp.sonar.host.url=https://sonar.elastic.dev
+systemProp.sonar.projectKey=elastic_logstash_AYm_nEbQaV3I-igkX1q9
```
[the remark this answers is not in the document]
Comment (pazone) [reply gradle.properties:11]: That's right. It's not a secret

**Comment by pazone** [comment]
@robbavey It's ready :)

File: ci/unit_tests.sh [hunk 3]
```diff
@@ -19,11 +19,19 @@ SELECTED_TEST_SUITE=$1
...
+    -Dsonar.pullrequest.base=$TARGET_BRANCH \
+    -Dsonar.scm.revision=$COMMIT_SHA \
+    --console=plain --warning-mode all
 elif [[ $SELECTED_TEST_SUITE == $"ruby" ]]; then
   echo "Running Ruby unit tests"
   ./gradlew rubyTests --console=plain --warning-mode all
 else
   echo "Running Java and Ruby unit tests"
-  ./gradlew test --console=plain
+  ./gradlew test jacocoTestReport sonar -Dsonar.token="${SONAR_TOKEN}" --console=plain --warning-mode all
```
Comment (robbavey) [inline ci/unit_tests.sh:36]: Our complex build system strikes again... 😩 

Currently this [fails](https://logstash-ci.elastic.co/job/elastic+logstash+main+multijob-unix-compatibility/os=amazon/412/) the Unix acceptance tests.

The Unix acceptance tests call `ci/unit_tests.sh` directly without a parameter, and also bypass the `ci/docker_unit_tests.sh` script.

This gives us a couple of problems - firstly, the system properties are not being set in this code path, so the sonar server cannot be found. Secondly, as we are bypassing the docker script, we're not setting the `ENV` variables.

I guess we could just call the `sonar` and `jacocoTestReport` tasks when we explicitly specify the language type... Or add a command line option to include/exclude running those tests, and update the jenkins jobs accordingly

**Review by robbavey — CHANGES_REQUESTED** [review]

File: ci/unit_tests.sh (hunk 3 above)
Comment (pazone) [reply ci/unit_tests.sh]: Updated the pr. The workaround should work now

**Review by robbavey — APPROVED** [review]
LGTM! 

Just one nit remaining about removing some commented code

**Pull request merged**